src/radical/pilot/raptor/master.py

```diff
@@ -261,84 +247,112 @@ def _state_cb(self, topic, msg):
+        master - the worker ranks are expected to syncronize their ranks as
+        needed.
+        '''
         with self._lock:
-            self._workers[uid] = dict()
-            self._workers[uid]['state'] = 'NEW'
+
+            tasks    = list()
+            base     = os.environ['RP_TASK_ID']
+
+            cfg            = copy.deepcopy(self._cfg)
+            cfg['descr']   = descr
+            cfg['info']    = self._info
+            cfg['ts_addr'] = self._task_service.addr
+
+            for i in range(count):
+
+                uid        = '%s.worker.%04d' % (base, i)
+                cfg['uid'] = uid
+
+                fname = './%s.json' % uid
+                ru.write_json(cfg, fname)
+
+                td = dict()
+                td['named_env']        = descr.get('named_env')
+                td['cpu_processes']    = descr['cpu_processes']
+                td['cpu_process_type'] = 'MPI'
+                td['cpu_thread_type']  = 'POSIX'
+                td['cpu_threads']      = descr.get('cpu_threads', 1)
+                td['gpu_processes']    = descr.get('gpu_processes', 0)
+                td['environment']      = descr.get('environment', {})
+
+                # this master is obviously running in a suitable python3 env,
+                # so we expect that the same env is also suitable for the worker
+                # NOTE: shell escaping is a bit tricky here - careful on change!
+                td['executable'] = 'python3'
+                td['arguments']  = [
+                        '-c',
+                        'import radical.pilot as rp; '
+                        "rp.raptor.Worker.run('%s', '%s', '%s')"
+                            % (descr.get('worker_file', ''),
+                               descr.get('worker_class', 'DefaultWorker'),
+                               fname)]
+
+
+
+                # all workers run in the same sandbox as the master
+                sbox = os.environ['RP_TASK_SANDBOX']
+                task = dict()
+
+                task['description']       = TaskDescription(td).as_dict()
+                task['state']             = rps.AGENT_STAGING_INPUT_PENDING
+                task['status']            = 'NEW'
+                task['type']              = 'task'
+                task['uid']               = uid
+                task['task_sandbox_path'] = sbox
+                task['task_sandbox']      = 'file://localhost/' + sbox
+                task['pilot_sandbox']     = os.environ['RP_PILOT_SANDBOX']
+                task['session_sandbox']   = os.environ['RP_SESSION_SANDBOX']
+                task['resource_sandbox']  = os.environ['RP_RESOURCE_SANDBOX']
+                task['pilot']             = os.environ['RP_PILOT_ID']
+                task['resources']         = {'cpu': td['cpu_processes'] *
+                                                    td.get('cpu_threads', 1),
+                                             'gpu': td['gpu_processes']}
```

Comment:
might be a little off-topic - do we actually use this info (`task['resources']`) or it is only for records?

also, shouldn't it be `'gpu': td['cpu_processes'] * td['gpu_processes']`?

p.s. similar setup is in `agent.scheduler` as well and I would have the same questions..

Reply:
It's used by analytics only

Reply:
It's used by analytics only

Reply:
TODO AM: fix GPU needs to be multiplied bu CPU processes!

Reply:
TODO AM: fix GPU needs to be multiplied bu CPU processes!